srcStatic/Log.cpp

```diff
@@ -0,0 +1,28 @@
+#include "Log.h"
+#include "op2ext-Internal.h"
+
+
+// Use anonymous namespace to keep globals private
+// Prevents other files from gaining access with a sneaky extern declaration
+namespace {
+  // Static (constant and zero) initialization happens before dynamic initialization
+  // Such values are well defined if accessed by a global object constructor before Main/DllMain starts
+  // This pointer is safely set to null before other globals are initialized
+  Logger* logger = nullptr;
```

Comment:
Why did you choose to use a 'raw' pointer here instead of a unique or shared pointer?

Should we ensure the logger is deleted before closing out of the application?

Thanks,
Brett

Reply:
I documented that responsibility was for the caller. Hence no ownership transfer takes place. This makes it safe to pass local or global objects. I used this is my GMock attempt, before I realized I had an older version installed which didn't work with the syntax I wanted. The test case would have looked like:
```
TEST(Log, SetLoggerReceivesMessages) {
	const auto message = std::string("Logger should receive this");

	MockLogger logger;
	EXPECT_CALL(logger, Log(message));

	EXPECT_NO_THROW(SetLogger(&logger));
	EXPECT_NO_THROW(Log(message));
	EXPECT_NO_THROW(SetLogger(nullptr));
}
```

The `logger` instance is local and would have been cleaned up automatically when the test function ended. If I'd put `unique_ptr` into the method signature, there would have been no safe way to use a local or global object. I would have been forced to wrap it in a `unique_ptr`, which would have forced me to use dynamic allocation.

The caller is of course free to use `unique_ptr` if they want.
```
unique_ptr<Logger> logger;

void CreateLogger() {
  logger = make_unique<LoggerFile>();
  SetLogger(logger.Get());
} // No automatic cleanup
```

In this case, when the application ends, the runtime will ensure cleanup of the `Logger` object stored in the global `unique_ptr`.

I could also have use a global:
```
LoggerFile logger;

void Init() {
  SetLogger(&logger);
  // ...
}
```

Again, this object is automatically destructed when the application terminates.

----

By keeping the `unique_ptr` out of the interface it gives the user the choice of how they want to manage object lifetime, and gives them the flexibility to use objects that are either dynamically allocated, allocated on the stack, or global.